### DISCUSSION TITLE
Tools 📦

### DISCUSSION THREAD

**Discussion by mediumhust** [TinyHust/notes#237] [Vault]
- [https://retool.com](https://retool.com/)
- [https://dynamitejobs.com](https://dynamitejobs.com/)
- [https://penpot.app](https://penpot.app/)
- [https://github.com/trpc/trpc](https://github.com/trpc/trpc)
- [https://github.com/colinhacks/zod](https://github.com/colinhacks/zod)
- [https://github.com/ianstormtaylor/superstruct](https://github.com/ianstormtaylor/superstruct)
- https://designer.microsoft.com/
- https://app.clipchamp.com/
- https://www.npmjs.com/package/config
- https://swc.rs/
- [huggingface spaces](https://huggingface.co/docs/hub/spaces-overview)
- [resend](https://resend.com/) - free SMTP

**Comment by mediumhust** [comment]
## Media
- [https://beta.elevenlabs.io/](https://beta.elevenlabs.io/)
- [https://imgupscaler.com](https://imgupscaler.com/)
- [https://www.y2mate.com/en569](https://www.y2mate.com/en569)
- [https://www.capcut.com/](https://www.capcut.com/)

**Reply by mediumhust** [reply]
## Text to speech
- https://elevenlabs.io/
- https://www.kreadoai.com/

**Reply by mediumhust** [reply]
## Animate character using audio
- https://www.adobe.com/express/create/video/animation

**Reply by mediumhust** [reply]
## Voice, Video, SMS
- https://www.sinch.com/

**Comment by mediumhust** [comment]
## AWS
- [Hands-on](https://cloudjourney.awsstudygroup.com/)
- [workshops](https://workshops.aws/)
- [Cloud Development Kit (CDK) libraries](https://constructs.dev/)
- [CloudWatch Logs Insights query](https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/CWL_QuerySyntax.html)
- [Policy Generator](https://awspolicygen.s3.amazonaws.com/policygen.html)
- Lambda
  - Lambda@Edge
- [Federating from Microsoft Azure AD to AWS](https://000041.awsstudygroup.com/2-prerequiste/)
- [Assume Role](https://www.youtube.com/watch?v=y4V2XK1Vdhc&ab_channel=VietTran)
- DataSync

**Reply by mediumhust** [reply]
### S3
- [select object](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3/client/select_object_content.html)
- [cli multipart upload](https://repost.aws/knowledge-center/s3-multipart-upload-cli)
- [lifecycle](https://docs.aws.amazon.com/AmazonS3/latest/userguide/lifecycle-transition-general-considerations.html)
- [transfer acceleration](https://aws.amazon.com/s3/transfer-acceleration/)
- [replication](https://docs.aws.amazon.com/AmazonS3/latest/userguide/replication-walkthrough-2.html)

**Reply by mediumhust** [reply]
### [CloudFormation](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/Welcome.html)
- [mappings](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/mappings-section-structure.html)
- [parameters](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/parameters-section-structure.html)
- [sample](https://github.com/mediumhust/cfn101)
- [Lambda@Edge](https://viblo.asia/p/terraform-series-bai-12-terraform-ab-testing-deployment-1Je5EAWj5nL)
  - https://github.com/mediumhust/terraform-series/tree/main/bai-11

**Reply by mediumhust** [reply]
### [EFS](https://aws.amazon.com/efs/)
- [amazon-efs-utils](https://docs.aws.amazon.com/efs/latest/ug/installing-amazon-efs-utils.html)
- [botocore](https://docs.aws.amazon.com/efs/latest/ug/install-botocore.html)
- enable VPC DNS hostname
- security group allow NFS for ec2 security group

**Reply by mediumhust** [reply]
### EC2
- Bastion Host
```
# forwarding authentication
ssh-add -K key.pem
# check
# ssh-add -l
ssh -A user@ip
# ssh-add -l
ssh -A user2@privateIP
```
- install stress on amazon linux 2
```
sudo amazon-linux-extras install epel -y
sudo yum install stress -y

#Run stress for 5mins
sudo stress --cpu 4 --timeout 320
```

**Reply by mediumhust** [reply]
### ELB
#### ALB
- Stickiness
- Deregistration Delay (Connection Draining)
#### NLB

**Reply by mediumhust** [reply]
### SSM
- Parameter Store

**Reply by mediumhust** [reply]
### Direct Connect
- [ref](https://www.youtube.com/watch?v=pMwFtm3N2Ls)

**Reply by mediumhust** [reply]
### EventBridge

**Reply by mediumhust** [reply]
### Cloud9

**Reply by mediumhust** [reply]
### IAM
- PassRole
- Identity-based Policy
- Resource-based Policy
- Permission Boundary
- [Assume Role](https://www.youtube.com/watch?v=y4V2XK1Vdhc&ab_channel=VietTran)

**Reply by mediumhust** [reply]
### [CodeBuild](https://github.com/mediumhust/simple-app)
### CodeDeploy
- Blue/Green Deployments

**Reply by mediumhust** [reply]
### RAM (Resource Access Manager)

**Reply by mediumhust** [reply]
### VPC
- Transit Gateway
  - Single Route Table
  - Multiple Route Table
  - Peering Attachment
- VPC endpoint
- Client VPN
  - https://dev.to/kasukur/how-to-setup-an-aws-client-vpn-51f

**Reply by mediumhust** [reply]
### KMS
- CMK

**Reply by mediumhust** [reply]
### Config

**Reply by mediumhust** [reply]
### Step Functions
- [example](https://github.com/mjzone/sfn-crash-course)
- [Serverless Step Functions](https://www.serverless.com/plugins/serverless-step-functions)

**Reply by mediumhust** [reply]
### API Gateway
- [whitelist ip](https://repost.aws/knowledge-center/api-gateway-resource-policy-access)

**Reply by mediumhust** [reply]
## Alternatives 
### All: [localstack](https://github.com/localstack/localstack)
### S3: MinIO
- https://min.io/
- https://www.npmjs.com/package/nestjs-minio-client
- https://min.io/docs/minio/linux/developers/javascript/API.html

**Reply by mediumhust** [reply]
### WAF

**Reply by mediumhust** [reply]
### SAM
- https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-getting-started.html
- https://github.com/mediumhust/aws-sam-typescript-layers-example

**Reply by mediumhust** [reply]
## Data Engineering

### EMR
- https://www.youtube.com/watch?v=ZFns7fvBCH4

### MSF (Managed Service for Apache Flink)
- https://www.youtube.com/watch?v=zfK3mJ_bkSc

**Reply by mediumhust** [reply]
### EKS
- Container Insights with CloudWatch Observability EKS add-on
  -  https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/Container-Insights-setup-EKS-addon.html
  - https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/install-CloudWatch-Observability-EKS-addon.html

**Reply by mediumhust** [reply]
### [MCP Servers](https://awslabs.github.io/mcp/)

**Comment by mediumhust** [comment]
## Service
- [railway.app](https://railway.app/)
- https://amplication.com/
- [singlestore](https://www.singlestore.com/)

**Reply by mediumhust** [reply]
### Automation
- https://www.make.com/en
- https://n8n.io/

**Comment by mediumhust** [comment]
## Nodejs file libs
- [csv-writer](https://www.npmjs.com/package/csv-writer)
- [ssh2-sftp-client](https://www.npmjs.com/package/ssh2-sftp-client)
  - https://loclv.hashnode.dev/nodejs-typescript-ssh2-sftp-client-package-phan-2-fast-download-upload
- [basic-ftp](https://github.com/patrickjuchli/basic-ftp)
  - https://www.npmjs.com/package/nestjs-ftp
- [xml-js](https://www.npmjs.com/package/xml-js)
- [jszip](https://www.npmjs.com/package/jszip)

**Comment by mediumhust** [comment]
## Framworks:
- [adminjs](https://docs.adminjs.co)
- [payloadcms](https://payloadcms.com/)
- https://adonisjs.com/

**Comment by mediumhust** [comment]
## K8s
- [eksctl](https://eksctl.io/)
- [skaffold](https://skaffold.dev/)
- [kind](https://kind.sigs.k8s.io/)
- [minikube](https://minikube.sigs.k8s.io/docs/start/)
- [istio](https://istio.io/latest/docs/setup/getting-started/)
- [grafana](https://grafana.com/)
- [prometheus](https://prometheus.io/)
- [argo](https://argoproj.github.io/)
  - https://dev.to/thenjdevopsguy/registering-a-new-cluster-with-argocd-12mn
  - [Argo Rollout](https://argoproj.github.io/rollouts/)
    - [canary](https://viblo.asia/p/chien-luoc-trien-khai-ung-dung-canary-khong-he-kho-voi-argo-rollout-0gdJzxAnVz5)
    - [blue green](https://viblo.asia/p/chien-luoc-trien-khai-bluegreen-tren-kubernetes-thuc-hien-the-nao-yZjJY956JOE)
- [fluxcd](https://fluxcd.io/)
  - [flagger](https://flagger.app/)
- [vcluster](https://www.vcluster.com/)
- [helm](https://helm.sh/)
  - [helmfile](https://helmfile.readthedocs.io/en/latest/)
- ingress controller
  - [kong](https://konghq.com/products/kong-ingress-controller)
  - [ambassador](https://www.getambassador.io/) (AES)
- [kopf](https://github.com/nolar/kopf)
- Admission Controllers
  - Kyverno
    - Kyverno CLI
  - [OPA Gatekeeper](https://github.com/open-policy-agent/gatekeeper)
    - [Conftest](https://www.conftest.dev/)
- [Reloader](https://github.com/stakater/Reloader)
- [KubeVPN](https://github.com/KubeNetworks/kubevpn)
  - https://viblo.asia/p/kubevpn-cong-cu-giup-debug-tren-moi-truong-kubernetes-nhanh-chong-va-hieu-qua-hon-AZoJjKbELY7

**Reply by mediumhust** [reply]
### Base
- [configmap](https://viblo.asia/p/kubernetes-series-bai-8-configmap-and-secret-truyen-cau-hinh-vao-container-RQqKL6rrl7z)
- [crd](https://viblo.asia/p/kubernetes-series-bai-19-adding-custom-resource-to-kubernetes-end-eW65GB8alDO)
- [downward-api](https://viblo.asia/p/kubernetes-series-bai-10-downward-api-truy-cap-pod-metadata-m68Z0eGdlkG)
- [taints and tolerations](https://viblo.asia/p/kubernetes-series-bai-17-advanced-scheduling-taints-and-tolerations-924lJBDalPM)
- [node affinity and pod affinity](https://viblo.asia/p/kubernetes-series-bai-18-advanced-scheduling-node-affinity-and-pod-affinity-gAm5y7jqZdb)

**Reply by mediumhust** [reply]
### REFs
- [workshop](https://archive.eksworkshop.com/)
- [argocd](https://viblo.asia/p/k8s-phan-21-xay-dung-luong-gitops-voi-jenkins-va-argocd-y3RL1aX7Lao)
- [kubernetes-client JS](https://github.com/kubernetes-client/javascript)
- [vault](https://developer.hashicorp.com/vault/tutorials/kubernetes/kubernetes-sidecar#set-a-secret-in-vault)
  - [tls](https://developer.hashicorp.com/vault/tutorials/kubernetes/kubernetes-minikube-tls)

**Comment by mediumhust** [comment]
## AI
- https://poe.com
- https://runwayml.com
- https://remini.ai
- https://github.com/AUTOMATIC1111/stable-diffusion-webui
- https://civitai.com/
- https://huggingface.co/
- https://www.steve.ai/
- https://www.pika.art/
- https://lexica.art/
- https://claude.ai/
- https://github.com/joaomdmoura/crewAI
- Google Vertex AI Vision

**Reply by mediumhust** [reply]
- https://github.com/AUTOMATIC1111/stable-diffusion-webui
- https://civitai.com/

without nvidia card
```
git clone
pip install -r requiments.txt
python launch.py --skip-torch-cuda-test --precision full --no-half
```

**Reply by mediumhust** [reply]
- https://github.com/invoke-ai/InvokeAI

**Reply by mediumhust** [reply]
### API
- https://makersuite.google.com/app/apikey

**Reply by mediumhust** [reply]
### LLM
- [langchain](https://github.com/TinyHust/notes/discussions/413)
- [LlamaIndex](https://github.com/run-llama/llama_index)

**Comment by mediumhust** [comment]
## Cloud
- [terraform](https://registry.terraform.io/search/modules?page=2&provider=aws&q=aws)
- [Site-to-Site VPN Between GCP and AWS Cloud](https://viblo.asia/p/terraform-series-bai-16-multi-cloud-environment-oOVlY0ynZ8W)
- [pulumi](https://www.pulumi.com/)

**Reply by mediumhust** [reply]
## Terraform
- Base
- REFs
  - [example](https://github.com/mediumhust/terraform-series)
  - [https://github.com/terraform-in-action](https://github.com/terraform-in-action)
  - [vault](https://viblo.asia/p/terraform-series-bai-18-security-managing-secrets-with-vault-LzD5dooOljY)
  - eks terraform
    - https://github.com/mediumhust/aws-eks-example 
    - https://viblo.asia/p/toi-bat-dau-hoc-eks-7ymJX72R4kq

**Comment by mediumhust** [comment]
## NPM
- https://github.com/verdaccio/verdaccio

**Reply by mediumhust** [reply]
### monorepo
- [turborepo](https://turbo.build/)
  - https://github.com/mguay22/turborepo
- [nx](https://nx.dev/)
  - [module federation](https://nx.dev/recipes/angular/dynamic-module-federation-with-angular)

**Comment by mediumhust** [comment]
## Queue
- Kafka
- RabbitMQ 
- [pulsar](https://pulsar.apache.org/)
- [redpanda](https://redpanda.com/)
- [emqx](https://www.emqx.com/en) - Seamlessly connect, process, and stream real-time IoT data from edge to cloud—unlocking intelligent, actionable insights on a unified, scalable MQTT platform.

**Reply by mediumhust** [reply]
### Kafka 
- Kafka Web UI
  - [kafka-ui](https://docs.kafka-ui.provectus.io/configuration/authentication/aws-iam)
  - [kafdrop](https://github.com/obsidiandynamics/kafdrop)
- [MongoDB Kafka Connector](https://www.mongodb.com/docs/kafka-connector/current/)
- [ksqldb](https://ksqldb.io/)
- [KRaft](https://developer.confluent.io/learn/kraft/)

**Reply by mediumhust** [reply]
### Libs
- [Celery](https://docs.celeryq.dev/en/stable/)

**Comment by mediumhust** [comment]
## Database
- MongoDB
- PostgreSQL 
- [Cassandra](https://cassandra.apache.org/)
- [duckdb](https://duckdb.org/)
- [neo4j](https://neo4j.com/)
- [influxdb](https://github.com/influxdata/influxdb)
- [tidb](https://github.com/pingcap/tidb)
- [scylladb](https://www.scylladb.com/)
- vector database
  - [qdrant](https://qdrant.tech/)
  - [milvus](https://milvus.io/)
  - [chroma](https://www.trychroma.com/)
  - [pinecone](https://www.pinecone.io/)
- column-oriented (columnar)
  - [clickhouse](https://clickhouse.com/)
  - [redshift](https://aws.amazon.com/redshift/)

**Reply by mediumhust** [reply]
### Mongo
- [Database Triggers](https://www.mongodb.com/docs/atlas/app-services/triggers/database-triggers/)
- [Views](https://www.mongodb.com/docs/manual/core/views/)
- [Time Series](https://www.mongodb.com/docs/manual/core/timeseries-collections/)
- [MongoDB Kafka Connector](https://www.mongodb.com/docs/kafka-connector/current/)

**Reply by mediumhust** [reply]
### PostgreSQL
-  [Window Functions](https://www.postgresqltutorial.com/postgresql-window-function/)
- [Views](https://www.postgresqltutorial.com/postgresql-views/)
- [Triggers](https://www.postgresqltutorial.com/postgresql-triggers/)
- [json](https://www.digitalocean.com/community/tutorials/working-with-json-in-mysql)
- [postgresus](https://github.com/RostislavDugin/postgresus)

**Reply by mediumhust** [reply]
### Cassandra
- dbeaver: https://www.cloudcentric.dev/setting-up-cassandra-jdbc-driver-in-dbeaver-community-edition/

**Comment by mediumhust** [comment]
## Container
- Docker
- [Podman](https://podman.io/)

**Reply by mediumhust** [reply]
REFs:
- [Podman](https://viblo.asia/p/devops-podman-an-alternative-tool-for-docker-MkNLrZo7LgA)

**Reply by mediumhust** [reply]
### Private  registry
- [distribution](https://distribution.github.io/distribution/)
- [nexus](https://help.sonatype.com/en/sonatype-nexus-repository.html)
- [GitHub Packages](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry)
- [JFrog Artifactory CE](https://jfrog.com/community/open-source/)

**Comment by mediumhust** [comment]
## VSC extensions

- Auto Import
- Docker
- Kubernetes
- Doxygen Documentation Generator
- ESLint
- File Tree to Text Generator
- Git History
- Go
- Live Server
- Markdown All in One
- Prettier - Code formatter
- Remote - Containers
- Remote - SSH
- Remote - SSH: Editing Configuration Files
- REST Client
- solidity
- Todo Tree
- YAML
- Material Icon Theme
- Extension Pack for Java
- Spring Boot Extension Pack
- ES7+ React/Redux/React-Native snippets
- Prisma
- vscode-proto3
- aws toolkit
- Tailwind CSS IntelliSense
- Turntable
- [cline](https://github.com/cline/cline)
- [PostgreSQL](https://marketplace.visualstudio.com/items?itemName=ms-ossdata.vscode-pgsql)

**Comment by mediumhust** [comment]
## Machine Learning
REFs:
- [https://phamdinhkhanh.github.io/deepai-book/intro.html](https://phamdinhkhanh.github.io/deepai-book/intro.html)
- [https://machinelearningcoban.com/tabml_book/intro.html](https://machinelearningcoban.com/tabml_book/intro.html)
- https://d2l.aivivn.com/index.html
- https://machinelearningcoban.com/

**Reply by mediumhust** [reply]
### Libs
- [scikit-learn](https://github.com/scikit-learn/scikit-learn)
- [keras](https://keras.io/)
- [pytorch](https://github.com/pytorch/pytorch)

**Comment by mediumhust** [comment]
## Data Engineering
- [Apache Hadoop](https://hadoop.apache.org/)
- [Apache Spark](https://spark.apache.org/)
- [Apache Kafka](https://kafka.apache.org/)
- [Apache Airflow](https://airflow.apache.org/)
- [Apache Flink](https://flink.apache.org/)
- [Delta Lake](https://delta.io/)
- [Decodable](https://www.decodable.co/)
- ETL (Extract, Transform, Load)
  - [AWS Glue](https://aws.amazon.com/glue/)
  - [Google Cloud Dataflow](https://cloud.google.com/dataflow)
  - [Azure Data Factory](https://azure.microsoft.com/en-us/products/data-factory)
    - [Synapse Analytics](https://learn.microsoft.com/en-us/azure/synapse-analytics/)
  - [airbyte](https://airbyte.com/)
  - [databricks](https://www.databricks.com/)
  - [dagster](https://dagster.io/)
- CDC (Change Data Capture)
  - [Debezium](https://debezium.io/)
- Visualizations
  - [tableau](https://www.tableau.com/)
  - [power-bi](https://app.powerbi.com/)
  - [superset](https://superset.apache.org/)
  - [streamlit](https://github.com/streamlit/streamlit)
- [columnar data](https://en.wikipedia.org/wiki/Column-oriented_DBMS)
  - BigQuery
    - [DBT](https://www.getdbt.com/) (Data Build Tool)
  - Redshift
- Unit Tests
  - [deequ](https://github.com/awslabs/deequ)

**Reply by mediumhust** [reply]
### Tableau
- https://www.youtube.com/watch?v=K3pXnbniUcM
- https://www.datawithbaraa.com/tableau/tableau-thank-you/

**Reply by mediumhust** [reply]
REFs:
- https://github.com/mediumhust/Japan-visa-data-engineering
- https://github.com/mediumhust/FootballDataEngineering
- https://github.com/mediumhust/YoutubeAnalytics
- https://github.com/mediumhust/dbt-bigquery-crash-course
- https://github.com/mediumhust/e2e-data-engineering
- https://github.com/mediumhust/SparkingFlow
- https://github.com/mediumhust/RealtimeStreamingEngineering
- https://github.com/mediumhust/Kubernetes-For-DataEngineering
- https://github.com/mediumhust/RedditDataEngineering
- https://github.com/mediumhust/EMR-for-data-engineers
- https://github.com/mediumhust/changecapture-e2e
- https://github.com/mediumhust/realtime-voting-data-engineering
- https://github.com/mediumhust/pipeline

**Reply by mediumhust** [reply]
Sample data
- https://www.kaggle.com/datasets

**Reply by mediumhust** [reply]
### Spark
spark-submit
```
docker exec -it spark-master-continer-name spark-submit --master spark://172.18.0.4:7077 --packages com.datastax.spark:spark-cassandra-connector_2.13:3.4.1,org.apache.spark:spark-sql-kafka-0-10_2.12:3.4. jobs/spark/spark-consumer.py
```

- https://github.com/steveloughran/winutils

**Reply by mediumhust** [reply]
### [Databricks Community Edition](https://community.cloud.databricks.com/)

**Reply by mediumhust** [reply]
### Airflow
#### REFs:
- https://github.com/mediumhust/data-engineer-zoomcamp-project
- https://viblo.asia/p/data-science-dung-apache-airflow-phien-ban-cuc-nhe-o-local-voi-docker-compose-x7Z4DAjPJnX

**Comment by mediumhust** [comment]
## CMD
- [linux command explain](https://explainshell.com/explain?cmd=k+get+pods+%7C+grep+Pending+%7C+wc+-l)
- https://www.tecmint.com/page/2/?s=ls+command

**Comment by mediumhust** [comment]
## Tracing, Metrics, Logs, Monitoring, Analytics
- https://newrelic.com/
- [Jaeger](https://www.jaegertracing.io/)
- https://opentelemetry.io/
- prometheus
- [aspecto](https://www.aspecto.io/)
- [zipkin](https://zipkin.io/)

**Reply by mediumhust** [reply]
### [Open Telemetry](https://opentelemetry.io/) (Otel)
- nestjs opentelemetry
  - https://medium.com/@alexanderknips/open-telemetry-with-multiple-nestjs-and-logs-in-tracing-62453e3e6cb2
  - https://github.com/mediumhust/nest-open-telemetry
  - https://github.com/mediumhust/otel-workshop
  - https://logz.io/blog/nestjs-javascript-opentelemetry-auto-instrumentation/
  - https://www.aspecto.io/blog/opentelemetry-collector-guide/

**Reply by mediumhust** [reply]
### Prometheus & Grafana
- https://www.youtube.com/watch?v=2ESOGJTXv1s
- https://github.com/mediumhust/nestjs-metrics
- https://github.com/Bhoopesh123/OpenTelemetry

**Reply by mediumhust** [reply]
### Monitoring
- [uptime-kuma](https://github.com/louislam/uptime-kuma)
  - https://thuanbui.me/uptime-kuma/
- 3rd services
  - [hetrixtools](https://hetrixtools.com/)
  - [uptimerobot](https://uptimerobot.com/)

**Reply by mediumhust** [reply]
### Analytics
- [sentry](https://sentry.io/)
- [hotjar](https://www.hotjar.com/)

**Reply by mediumhust** [reply]
### Logs
- [logfire](https://pydantic.dev/)

**Reply by mediumhust** [reply]
### Misc
- log rotation: [logrotate](https://betterstack.com/community/guides/logging/how-to-manage-log-files-with-logrotate-on-ubuntu-20-04/)
- [betterstack](https://betterstack.com/)

**Comment by mediumhust** [comment]
## Canvas
- ### 2D
  - https://createjs.com/
  - [Egret Engine](https://github.com/egret-labs/egret-core)
- ### 3D
  - ThreeJs 
- ### Game engine
  - [Egret Engine](https://github.com/egret-labs/egret-core)
  - [LayaAir](https://github.com/layabox/LayaAir)
  - https://phaser.io/

**Comment by mediumhust** [comment]
## Payment
- VN
  - https://docs.sepay.vn/

**Comment by mediumhust** [comment]
## DevTool
- https://github.com/theajack/disable-devtool

**Comment by mediumhust** [comment]
## Game
- private: https://bbs.90gm.vip/

**Comment by mediumhust** [comment]
### Swagger/OpenAPI
- [scalar](https://github.com/scalar/scalar)
- [redoc](https://github.com/Redocly/redoc)

**Comment by mediumhust** [comment]
## Jupyter Notebook
- https://colab.google/
- https://www.kaggle.com/code

**Reply by mediumhust** [reply]
### Colab
check CUDA version
```
!nvidia-smi
# or
!nvcc --version
```

**Comment by mediumhust** [comment]
## Misc

### diagram
- https://www.eraser.io/
- https://cloudairy.com/
- https://excalidraw.com/

### folder tree
- https://tree.nathanfriend.io/
- https://ascii-tree-generator.com/

### document management system
- [paperless-ngx](https://docs.paperless-ngx.com/)
  - https://thuanbui.me/paperless-ngx/
  - [paperless-ai](https://github.com/clusterzx/paperless-ai)
    - https://www.youtube.com/watch?v=ygtp8Z5Mslk

### [chatbot-ui](https://github.com/mckaywrigley/chatbot-ui)

### sso
- [authelia](https://www.authelia.com/)
- [clerk](https://clerk.com/)

### cPanel
- [OpenLiteSpeed](https://docs.litespeedtech.com/cloud/docker/ols-wordpress/)

### reverse proxy
- [Caddy](https://caddyserver.com/)
  - https://thuanbui.me/caddy-server/
- [Nginx Proxy Manager(NPM)](https://nginxproxymanager.com/guide/)
- [traefik](https://traefik.io/traefik/)
- [envoy](https://www.envoyproxy.io/)

### note taking app
- [hedgedoc](https://github.com/hedgedoc/hedgedoc)
- [joplin](https://github.com/laurent22/joplin)
- [memos](https://github.com/usememos/memos)

### docker/k8s ui
- [portainer](https://www.portainer.io/)

### homeserver
- [Cosmos](https://cosmos-cloud.io/)
- [runtipi](https://github.com/runtipi/runtipi)

### generate random text file
- https://onlinefiletools.com/generate-random-text-file
- https://generate-random.org/encryption-key-generator

### ssh client
- [mobaxterm](https://mobaxterm.mobatek.net/)
- [termius](https://termius.com/)

### test
- [jmeter](https://jmeter.apache.org/)
- [playwright](https://playwright.dev/)
- [vegeta](https://github.com/tsenart/vegeta)

### prompt theme
- [ohmyposh](https://ohmyposh.dev/)
- [fishshell](https://fishshell.com/)

### latex
- https://en.wikibooks.org/wiki/LaTeX/Mathematics
- [mathpix](https://mathpix.com/)
- https://latexeditor.lagrida.com/
- https://tug.ctan.org/info/undergradmath/undergradmath.pdf
- https://quickref.me/latex.html

### low-code automation
- [retool](https://retool.com/)
- [windmill](https://github.com/windmill-labs/windmill)
- [appsmith](https://www.appsmith.com/)

### vpn
- [pangolin](https://pangolin.net/)

**Reply by mediumhust** [reply]
### database
- [appwrite](https://appwrite.io/)
- [neondb](https://neon.tech/)
- [retool](https://retool.com/)

**Reply by TinyHust** [reply]
### math
- [mathworld](https://mathworld.wolfram.com/)

**Reply by mediumhust** [reply]
### collaborative wiki
- https://github.com/docmost/docmost (alternative to Confluence and Notion)

**Reply by mediumhust** [reply]
### mock api
- [mockapi](https://mockapi.io/)

**Reply by mediumhust** [reply]
### orchestration framework
- [windmill](https://github.com/windmill-labs/windmill)
  - [competitors](https://www.windmill.dev/docs/compared_to/peers)
- [kestra](https://kestra.io/)
- [prefect](https://www.prefect.io/)
- [retool](https://retool.com/)

**Comment by mediumhust** [comment]
## Nginx
- [Nginx Proxy Manager](https://nginxproxymanager.com/guide/)
  - https://thuanbui.me/nginx-proxy-manager/

**Comment by mediumhust** [comment]
## UI Kit
- [tabler](https://tabler.io/admin-template)
- https://ui.aceternity.com/
- https://ui.shadcn.com/

**Comment by mediumhust** [comment]
## Network, proxy, vpn
- cloudflare
  - [auto update DNS record](https://thuanbui.me/cap-nhat-ip-dong-cho-ten-mien-qua-cloudflare-de-truy-cap-homelab-tai-nha/)
- WOL
  - [upsnap](https://vnrom.net/2023/12/upsnap/)
  - https://github.com/seriousm4x/UpSnap
- vpn
  - https://pangolin.net/
- tunnel
  - [cloudflared](https://github.com/cloudflare/cloudflared)

**Comment by mediumhust** [comment]
## GCP

**Reply by mediumhust** [reply]
### Firebase

**Reply by mediumhust** [reply]
### Alternatives
- **Firebase**: 
  - supabase
    - https://supabase.com/docs/guides/self-hosting/docker
    - https://supabase.com/docs/guides/cli/getting-started?queryGroups=platform&platform=linux
  - [appwrite](https://appwrite.io/docs/advanced/self-hosting)

**Comment by mediumhust** [comment]
## Git
### git hook
- [husky](https://github.com/typicode/husky)
- [lefthook](https://github.com/evilmartians/lefthook)

### git action
- Self host
  - [RunsOn](https://runs-on.com/)
  - [AWS Codebuild](https://docs.aws.amazon.com/codebuild/latest/userguide/sample-github-action-runners.html)

### misc
- [gitleaks](https://github.com/gitleaks/gitleaks)

**Comment by TinyHust** [comment]
## [Computer Vision](https://github.com/TinyHust/notes/discussions/440#discussion-7477048)

**Comment by mediumhust** [comment]
## [Alternative Services](https://github.com/TinyHust/notes/discussions/451)

**Comment by mediumhust** [comment]
## uncategory
- [meilisearch](https://www.meilisearch.com/)
- [ffmpeg](https://ffmpeg.org/)
  - [docker image](https://hub.docker.com/r/linuxserver/ffmpeg)
- [s5cmd](https://github.com/peak/s5cmd)
  - [docker image](https://hub.docker.com/r/peakcom/s5cmd)
- [jqlang](https://jqlang.org/)
